File: TEP_2017_2/UVA/10689.c
#include <stdio.h>

int dp[15000];

int main(){
	int cases = 0;		
	int a, b, n, m;

	int exps[] = {1, 10, 100, 1000, 10000};

	scanf("%d", &cases);

	int i;
	for(i = 0; i < cases; ++i){
		scanf("%d %d %d %d", &a, &b, &n, &m);	

		dp[0] = a % exps[m];
		dp[1] = b % exps[m];

		int j;
		for(j = 2; j < 15000; ++j){
			int tmp = a + b;	
			a = b;
			b = tmp % exps[m];

			dp[j] = b;
		}

		printf("%d\n", dp[n % 15000]);
	}

	return 0;
}

File: TEP_2017_2/UVA/10689.py
# TLE

from sys import stdin

dp = [0] * 15000

count = 0
for line in stdin:
    line = line.strip()

    if count:
        a, b, n, m = [int(i) for i in line.split()]

        if not n:
            print(a)
            continue
        elif n == 1:
            print(b)
            continue
        
        for i in range(1, 15000):
            tmp = a + b
            a = b
            b = tmp%10**m
            dp[i - 1] = b

        print(dp[n % 15000 - 2])

    else:
        cases = line
        count += 1
